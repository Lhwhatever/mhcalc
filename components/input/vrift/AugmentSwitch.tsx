import { Avatar, Checkbox, makeStyles, Typography } from '@material-ui/core'
import React from 'react'
import { textToKebabCase } from '../../../utils/str'
import { InputChangeEvent } from '../types'

export interface AugmentSwitchProps {
    label: string
    id?: string
    iconUri: string
    checked?: boolean
    onChange: (state: boolean) => void
}

const useStyles = makeStyles({
    augmentIcon: {
        width: 30,
        height: 30
    }
})

const AugmentSwitch = (props: AugmentSwitchProps): JSX.Element => {
    const { label, id, iconUri, checked, onChange } = props
    const labelId = id ?? 'augment-switch-' + textToKebabCase(label)

    const classes = useStyles()

    const handleCheckboxChange = (event: InputChangeEvent) => {
        onChange(event.target.checked)
    }

    return (
        <React.Fragment>
            <Checkbox inputProps={{ 'aria-labelledby': labelId }} checked={checked} onChange={handleCheckboxChange} />
            <Avatar src={iconUri} className={classes.augmentIcon} />
            <Typography variant="body1" id={labelId}>
                {label}
            </Typography>
        </React.Fragment>
    )
}

export default AugmentSwitch
