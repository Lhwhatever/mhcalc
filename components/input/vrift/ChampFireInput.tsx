import React from 'react'
import { Switch, FormControlLabel } from '@material-ui/core'
import { InputChangeEvent } from '../types'

export interface ChampFireInputProps {
    value?: boolean
    onChange?: (state: boolean) => void
}

const ChampFireInput = (props: ChampFireInputProps): JSX.Element => {
    const { value, onChange = () => undefined } = props

    const handleChange = (event: InputChangeEvent) => {
        onChange(event.target.checked)
    }

    return (
        <FormControlLabel
            control={<Switch checked={value} onChange={handleChange} inputProps={{ role: 'switch checkbox' }} />}
            label="Champion's Fire"
        />
    )
}

export default ChampFireInput
