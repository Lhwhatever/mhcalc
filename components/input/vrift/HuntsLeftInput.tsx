import { TextField } from '@material-ui/core'
import React from 'react'
import { InputEvent } from '../types'

export interface HuntsLeftInputProps {
    fullWidth?: boolean
    value?: number
    onChange?: (event: InputEvent) => void
}

const HuntsLeftInput = (props: HuntsLeftInputProps): JSX.Element => {
    return <TextField label="Hunts Left" type="number" variant="outlined" {...props} />
}

export default HuntsLeftInput
