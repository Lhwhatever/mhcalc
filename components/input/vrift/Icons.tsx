import { Avatar, withStyles } from '@material-ui/core'
import React from 'react'

export const SyncIcon = (): JSX.Element => (
    <Avatar
        alt="Sync"
        src="https://www.mousehuntgame.com/images/ui/hud/rift_valour/power_up_stamina.png"
        variant="rounded"
    />
)

export const SpeedIcon = (): JSX.Element => (
    <Avatar
        alt="Speed"
        src="https://www.mousehuntgame.com/images/ui/hud/rift_valour/power_up_long_stride.png"
        variant="rounded"
    />
)

export const SiphonIcon = (): JSX.Element => (
    <Avatar
        alt="Siphon"
        src="https://www.mousehuntgame.com/images/ui/hud/rift_valour/power_up_siphon.png"
        variant="rounded"
    />
)

export interface ChampFireIconProps {
    state?: boolean
}

const champFireStyles = {
    default: {
        width: '100%',
        paddingBottom: '100%',
        backgroundImage: 'url(https://www.mousehuntgame.com/images/ui/hud/rift_valour/fuel.png)',
        backgroundPosition: '0 0',
        backgroundSize: '100%',
        borderRadius: '50%'
    },
    active: {
        backgroundPosition: '0% 100%'
    }
}

export const ChampFireIcon = withStyles(champFireStyles)(
    ({ state = false }: ChampFireIconProps): JSX.Element => (
        <div
            style={{
                ...champFireStyles.default,
                ...(state ? champFireStyles.active : undefined)
            }}
            role="img"
            aria-label="champion's fire"
        />
    )
)
