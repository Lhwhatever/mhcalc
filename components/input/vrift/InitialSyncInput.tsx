import { MenuItem, TextField, TextFieldProps } from '@material-ui/core'
import React from 'react'

export type InitialSync = 40 | 50 | 60 | 70 | 80 | 90 | 100

export type SyncInputProps = Omit<TextFieldProps, 'defaultValue'>

const InitialSyncInput = (props: SyncInputProps): JSX.Element => {
    return (
        <TextField select label="Sync" variant="outlined" {...props}>
            {Array(7)
                .fill(0)
                .map((_, i) => {
                    const sync = i * 10 + 40
                    return (
                        <MenuItem key={i} value={sync}>
                            Lvl. {i + 1} ({sync} Sync)
                        </MenuItem>
                    )
                })}
        </TextField>
    )
}

export default InitialSyncInput
