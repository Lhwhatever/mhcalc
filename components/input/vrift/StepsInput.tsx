import React from 'react'
import IntegerInput, { IntegerInputProps } from '../IntegerInput'

export type StepsInputProps = Omit<IntegerInputProps, 'min' | 'max'>

const StepsInput = (props: StepsInputProps): JSX.Element => {
    return <IntegerInput label="Steps" variant="outlined" min={0} {...props} />
}

export default StepsInput
