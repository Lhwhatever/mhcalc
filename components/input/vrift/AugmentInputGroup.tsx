import React from 'react'
import { Box, Typography, makeStyles } from '@material-ui/core'
import AugmentSwitch from './AugmentSwitch'
import { useSelector, useDispatch } from 'react-redux'
import { RootState } from '../../../redux/rootReducer'
import { Augments, updateAugment } from '../../../redux/ducks/vrift/simInput'

const useStyles = makeStyles((theme) => ({
    switchboxRoot: {
        display: 'grid',
        gridTemplateColumns: 'auto auto 1fr',
        gridColumnGap: theme.spacing(1),
        alignItems: 'center'
    }
}))

const AugmentInputGroup = (): JSX.Element => {
    const classes = useStyles()

    const enabledAugments = useSelector((state: RootState) => state.vrift.simInput.augments)
    const dispatch = useDispatch()

    const createAugmentStateHandler = (key: keyof Augments) => (state: boolean) => {
        dispatch(updateAugment({ target: key, state }))
    }

    return (
        <Box>
            <Typography variant="h4">Augments</Typography>
            <div className={classes.switchboxRoot}>
                <AugmentSwitch
                    label="Sigil Hunter"
                    iconUri="https://www.mousehuntgame.com/images/ui/hud/rift_valour/augments/highway_robbery.png"
                    checked={enabledAugments.sigilHunter ?? false}
                    onChange={createAugmentStateHandler('sigilHunter')}
                />
                <AugmentSwitch
                    label="Secret Research"
                    iconUri="https://www.mousehuntgame.com/images/ui/hud/rift_valour/augments/champion_pillage.png"
                    checked={enabledAugments.secretResearch ?? false}
                    onChange={createAugmentStateHandler('secretResearch')}
                />
                <AugmentSwitch
                    label="Super Siphon"
                    iconUri="https://www.mousehuntgame.com/images/ui/hud/rift_valour/augments/secret_passage.png"
                    checked={enabledAugments.superSiphon ?? false}
                    onChange={createAugmentStateHandler('superSiphon')}
                />
                <AugmentSwitch
                    label="Ultimate Umbra"
                    iconUri="https://www.mousehuntgame.com/images/ui/hud/rift_valour/augments/tower_ultimatum.png"
                    checked={enabledAugments.ultimateUmbra ?? false}
                    onChange={createAugmentStateHandler('ultimateUmbra')}
                />
                <AugmentSwitch
                    label="String Stepping"
                    iconUri="https://www.mousehuntgame.com/images/ui/hud/rift_valour/augments/string_stepping.png"
                    checked={enabledAugments.stringStepping ?? false}
                    onChange={createAugmentStateHandler('stringStepping')}
                />
            </div>
        </Box>
    )
}

export default React.memo(AugmentInputGroup)
