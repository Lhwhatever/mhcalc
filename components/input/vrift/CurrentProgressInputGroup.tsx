import { Box, Button, makeStyles, Tooltip, Typography } from '@material-ui/core'
import InfoIconOutlined from '@material-ui/icons/InfoOutlined'
import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { updateHuntsLeft, updateInitialSync, updateSteps } from '../../../redux/ducks/vrift/simInput'
import { RootState } from '../../../redux/rootReducer'
import { InputChangeEvent } from '../types'
import HuntsLeftInput from './HuntsLeftInput'
import { SyncIcon } from './Icons'
import InitialSyncInput, { InitialSync } from './InitialSyncInput'
import StepsInput from './StepsInput'

const useStyles = makeStyles((theme) => ({
    inputGrid: {
        display: 'grid',
        gridTemplateColumns: '1fr 3fr 3rem',
        gridGap: theme.spacing(1)
    },
    inputIconLabel: {
        gridColumn: '1 / span 1',
        display: 'flex',
        flexDirection: 'row-reverse',
        alignItems: 'center'
    },
    inputField: {
        gridColumn: '2 / span 1',
        '& .MuiTextField-root': {
            minWidth: 166
        }
    },
    inputAdornment: {
        gridColumn: '3 / span 1',
        display: 'flex',
        flexDirection: 'row',
        alignItems: 'center'
    }
}))

const CurrentProgressInputGroup = (): JSX.Element => {
    const classes = useStyles()

    const { initialSync, huntsLeft, steps } = useSelector((state: RootState) => state.vrift.simInput)
    const dispatch = useDispatch()

    const handleInitialSyncChange = (event: InputChangeEvent) => {
        dispatch(updateInitialSync(parseInt(event.target.value) as InitialSync))
    }

    const handleHuntsLeftChange = (huntsLeft: number | undefined) => {
        dispatch(updateHuntsLeft(huntsLeft))
    }

    const handleStepsChange = (steps: number | undefined) => {
        dispatch(updateSteps(steps))
    }

    const handleResetToStart = () => {
        handleHuntsLeftChange(initialSync)
        handleStepsChange(0)
    }

    return (
        <Box>
            <Typography variant="h4">Current Progress</Typography>
            <Box className={classes.inputGrid} mt={2}>
                <div className={classes.inputIconLabel}>
                    <SyncIcon />
                </div>
                <div className={classes.inputField}>
                    <InitialSyncInput
                        fullWidth
                        value={initialSync}
                        onChange={handleInitialSyncChange}
                        id="initial-sync-input"
                    />
                </div>
                <div className={classes.inputAdornment}>
                    <Tooltip title="If you are upgrading Sync during the run, use the final Sync level.">
                        <InfoIconOutlined />
                    </Tooltip>
                </div>
                <div className={classes.inputField}>
                    <HuntsLeftInput
                        fullWidth
                        value={huntsLeft}
                        onChange={handleHuntsLeftChange}
                        id="hunts-left-input"
                    />
                </div>
                <div className={classes.inputField}>
                    <StepsInput fullWidth value={steps} onChange={handleStepsChange} id="steps-input" />
                </div>
            </Box>
            <Box mt={2} display="flex" flexDirection="row-reverse">
                <Button id="btn-reset-run" variant="outlined" onClick={handleResetToStart}>
                    Reset Progress
                </Button>
            </Box>
        </Box>
    )
}

export default CurrentProgressInputGroup
