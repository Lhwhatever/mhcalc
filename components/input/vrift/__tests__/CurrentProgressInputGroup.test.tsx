import { Button, Tooltip } from '@material-ui/core'
import { createShallow } from '@material-ui/core/test-utils'
import InfoIconOutlined from '@material-ui/icons/InfoOutlined'
import { ShallowWrapper } from 'enzyme'
import React from 'react'
import { SimInputsState } from '../../../../redux/ducks/vrift/simInput'
import CurrentProgressInputGroup from '../CurrentProgressInputGroup'
import HuntsLeftInput from '../HuntsLeftInput'
import InitialSyncInput from '../InitialSyncInput'
import StepsInput from '../StepsInput'

jest.mock('react-redux', () => ({
    useDispatch: () => jest.fn(),
    useSelector: (): SimInputsState => ({
        initialSync: 100,
        huntsLeft: 100,
        steps: 0,
        augments: {}
    })
}))

describe('CurrentProgressInputGroup test', () => {
    let shallow: ReturnType<typeof createShallow>
    let wrapper: ShallowWrapper

    beforeAll(() => {
        shallow = createShallow()
    })

    beforeEach(() => {
        wrapper = shallow(<CurrentProgressInputGroup />)
    })

    it('should have a header', () => {
        expect(wrapper.text()).toMatch(/current progress/i)
    })

    it('should have the correct inputs', () => {
        expect(wrapper.find(InitialSyncInput)).toHaveLength(1)
        expect(wrapper.find(HuntsLeftInput)).toHaveLength(1)
        expect(wrapper.find(StepsInput)).toHaveLength(1)
    })

    it('should have a Tooltip > InfoOutlined', () => {
        const infoTooltip = wrapper.find(Tooltip).filterWhere((node) => node.contains(<InfoIconOutlined />))
        expect(infoTooltip).toHaveLength(1)
    })

    it('should have a button to reset progress', () => {
        const btn = wrapper.find(Button).filter('#btn-reset-run')
        expect(btn).toHaveLength(1)
        expect(btn.text()).toMatch(/reset progress/i)
    })

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })
})
