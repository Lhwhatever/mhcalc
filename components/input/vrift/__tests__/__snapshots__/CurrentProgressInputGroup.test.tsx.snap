// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentProgressInputGroup test should match snapshot 1`] = `
<Styled(MuiBox)>
  <WithStyles(ForwardRef(Typography))
    variant="h4"
  >
    Current Progress
  </WithStyles(ForwardRef(Typography))>
  <Styled(MuiBox)
    className="makeStyles-inputGrid-1"
    mt={2}
  >
    <div
      className="makeStyles-inputIconLabel-2"
    >
      <WithStyles(ForwardRef(Avatar))
        alt="Sync"
        src="https://www.mousehuntgame.com/images/ui/hud/rift_valour/power_up_stamina.png"
        variant="rounded"
      />
    </div>
    <div
      className="makeStyles-inputField-3"
    >
      <InitialSyncInput
        fullWidth={true}
        onChange={[Function]}
        value={100}
      />
    </div>
    <div
      className="makeStyles-inputAdornment-4"
    >
      <WithStyles(ForwardRef(Tooltip))
        title="If you are upgrading Sync during the run, use the final Sync level."
      >
        <InfoOutlinedIcon />
      </WithStyles(ForwardRef(Tooltip))>
    </div>
    <div
      className="makeStyles-inputField-3"
    >
      <HuntsLeftInput
        fullWidth={true}
        onChange={[Function]}
        value={100}
      />
    </div>
    <div
      className="makeStyles-inputField-3"
    >
      <StepsInput
        fullWidth={true}
        onChange={[Function]}
        value={0}
      />
    </div>
  </Styled(MuiBox)>
  <Styled(MuiBox)
    display="flex"
    flexDirection="row-reverse"
    mt={2}
  >
    <WithStyles(ForwardRef(Button))
      id="btn-reset-run"
      onClick={[Function]}
      variant="outlined"
    >
      Reset Progress
    </WithStyles(ForwardRef(Button))>
  </Styled(MuiBox)>
</Styled(MuiBox)>
`;
